### PR TITLE
iterative ref: build_ctf_array

### LINKED ISSUE #11: iterative ref: build_ctf
Decide on how ctf param info will be passed

Consider how to match particles and ctf_info

### PR DESCRIPTION
**Pull request by jedyeo**
Design choice that needs to be discussed!

We propose `self.ctf_info` to instead be a singular dictionary instead of a list of dicts. `ctf_info` would contain CTF config values, and contain an additional key, `ctf_params`, which holds 3 key-value pairs. The keys would be the defocus u, v, angles, and the values would be an ordered list (with each index corresponding to a particle).

```
self.ctf_info = {   
            'amplitude_contrast': 0.1,
            'b_factor': 0.0,
            'batch_size': len(self.particles),  # num samples
            'cs': 2.7,
            'ctf_size': 32,  # n_pix
            'kv': 300,
            'pixel_size': 3.2,
            'side_len': 32,
            'value_nyquist': 0.1,
            'ctf_params' : {   # user would also pass this part in
                'defocus_u' : [...],
                'defocus_v' : [...],
                'defocus_angle' : [...], 
            }
        }
```
   

This follows the required input/output from `linear_simulator.ctf`
https://github.com/compSPI/simSPI/blob/master/simSPI/linear_simulator/ctf.py

### CODE REVIEW

**Comment by jedyeo** [comment]
Looks like we can't import from `simSPI.linear_simulator` since it's not recognized as a module.

I've added a temporary fix in `environment.yml` to work around this for now, but ideally we can get a small `__init__.py` file merged https://github.com/compSPI/simSPI/pull/101

**Comment by jedyeo** [comment]
im going to rest now and get back to this tmrw morning :)

**Comment by geoffwoollard** [comment]
Let me look into `value_nyquist`

https://github.com/compSPI/simSPI/blob/master/simSPI/linear_simulator/ctf.py#L108

It's how far out (a ratio) in Fourier space. The edge of the box in Fourier space is "half Nyquist", which corresponds to double the space (of one pixel). 
https://en.wikipedia.org/wiki/Nyquist_rate

It's in units of Fourier pixels and controls the decay envelope of the CTF: 
https://github.com/compSPI/simSPI/blob/master/simSPI/linear_simulator/ctf.py#L160

```
        if self.config.b_factor == 0:
            decay = (
                np.sqrt(-np.log(self.config.value_nyquist))
                * 2.0
                * self.config.pixel_size
            )
            envelope = torch.exp(-self.frequency**2 * decay**2)
        else:
            envelope = torch.exp(-self.frequency**2 * self.config.b_factor / 4.0)
```

Keep `b_factor=0` so that you use `value_nyquist`. It's the value of the envelope at the edge of the box (at half-Nyquist, which people just sometimes call Nyquist for shorth and). You can see this from the equation above.

At the edge of the box (Half Nyquist ), frequency*pixel_size is 1/2 (that's why it's called half!). So `-self.frequency**2 * decay**2` is `(1/2)**2 * 2**2 * np.sqrt(-np.log(self.config.value_nyquist))**2` which is `-np.log(self.config.value_nyquist)`. Thus `envelope = +self.config.value_nyquist`. QED

**Comment by geoffwoollard** [comment]
Stick to the linear simulator ctf as much as possible, including how the ctf params are passed in. Match the ctf params dictionary in linear_simulator.ctf with how you take in the ctf, so that you don't have to write any parser.

If you don't write code, you don't have to test it!

**Comment by geoffwoollard** [comment]
> Design choice that needs to be discussed!
> 
> We propose `self.ctf_info` to instead be a singular dictionary instead of a list of dicts. `ctf_info` would contain CTF config values, and contain an additional key, `ctf_params`, which holds 3 key-value pairs. The keys would be the defocus u, v, angles, and the values would be an ordered list (with each index corresponding to a particle).
> 
> ```
> self.ctf_info = {   
>             'amplitude_contrast': 0.1,
>             'b_factor': 0.0,
>             'batch_size': len(self.particles),  # num samples
>             'cs': 2.7,
>             'ctf_size': 32,  # n_pix
>             'kv': 300,
>             'pixel_size': 3.2,
>             'side_len': 32,
>             'value_nyquist': 0.1,
>             'ctf_params' : {   # user would also pass this part in
>                 'defocus_u' : [...],
>                 'defocus_v' : [...],
>                 'defocus_angle' : [...], 
>             }
>         }
> ```
> 
> This follows the required input/output from `linear_simulator.ctf` https://github.com/compSPI/simSPI/blob/master/simSPI/linear_simulator/ctf.py

The "user" is whoever is calling this function. Could be a methods developer or practitioner. Hard to say now since it's early days. This function just reads the ctf params that come from somewhere else and doesn't change them. Don't use default values. In the tests, you'll have to put some values in, and those look like OK values. I would suggest also
* pixel_size: 128 to 256
* defocus_u,defocus_v: 0.5 to 2.5 since it's in um. have their ratio not too far from 1. e.g. 0.9 to 1.1
* angle_astigmatism: 0 to 2pi
* wavelength: 0.1968, (@jedyeo important!) which corresponds to 300 kev. Instead of the kv beam energy, linear_simulator.ctf uses the electron wavelength 
  * I think in units of Angstroms, from the docs and tests
  * https://github.com/compSPI/simSPI/blob/master/simSPI/linear_simulator/ctf.py#L50
  * https://github.com/compSPI/simSPI/blob/master/tests/test_ctf.py#L133
  *  1 A = 10 pm. See 300 kev in this table: https://www.jeol.co.jp/en/words/emterms/search_result.html?keyword=wavelength%20of%20electron

I'll also look into ctf_size != side_len. It could be related to zero padding...

**Comment by geoffwoollard** [comment]
@jedyeo @thisTyler @arjunsingh3600 @calhep you can play with this applet to get some intuition about the CTF
https://3dem.github.io/relion/ctf.html

Good pedagogical post
https://forum.scilifelab.se/t/thinking-about-ctf/114

**Comment by geoffwoollard** [comment]
You can look at the tests for how `ctf_params` is defined in linear_simulator and how you should therefore construct it in the tests. Take note of all the batches etc. Read up on the tests and docs of https://github.com/compSPI/simSPI/blob/master/simSPI/linear_simulator/ctf.py#L160

File: reconstructSPI/iterative_refinement/expectation_maximization.py [hunk 1]
```diff
@@ -23,9 +23,34 @@ class IterativeRefinement:
     particles : arr
         Particles to be reconstructed.
         Shape (n_particles, n_pix, n_pix)
-    ctf_info : list of dicts
-        Each dict contains CTF k,v pairs per particle.
-            Shape (n_particles,)
+    ctf_info : dict
```
Comment (geoffwoollard) [inline reconstructSPI/iterative_refinement/expectation_maximization.py:26]: I wouldn't repeat. Just say, see *** documentation and point to the place the ctf comes from. that way when it changes there it doesn't need to be updated here.

File: reconstructSPI/iterative_refinement/expectation_maximization.py [hunk 2]
```diff
@@ -372,15 +397,14 @@ def __init__(self, *args, **kwargs):
...
+        for k, v in self.ctf_info["ctf_params"].items():
             tensor_dict[k] = torch.from_numpy(v.reshape(tensor_shape))
 
         ctf_shape = (
             len(self.particles),
             self.particles.shape[1],
             self.particles.shape[2],
         )
-
         return ctf.get_ctf(tensor_dict).numpy().reshape(ctf_shape)
```
Comment (geoffwoollard) [inline reconstructSPI/iterative_refinement/expectation_maximization.py:385]: ok this parser looks good

File: reconstructSPI/iterative_refinement/expectation_maximization.py (hunk 1 above)
Comment (geoffwoollard) [reply reconstructSPI/iterative_refinement/expectation_maximization.py:26]: also the documentation is too restrictive. while these are reasonable/typical values for testing, others outside their range are still valid, even if a bit less common.

File: tests/test_expectation_maximization.py [hunk 3]
```diff
@@ -23,24 +23,40 @@ def n_particles():
...
-def test_ir(n_pix, n_particles):
+def rand_angle_list(n_particles):
+    """Get random astigmatism angle between 0 and 2pi."""
+    return np.random.uniform(low=0, high=2 * np.pi, size=(n_particles,))
+
+
+@pytest.fixture
+def rand_defocus(n_particles):
+    """Get random defocus values between 0.5 and 2.5."""
+    return np.random.uniform(low=0.5, high=2.5, size=(n_particles,))
```
Comment (geoffwoollard) [inline tests/test_expectation_maximization.py:34]: good for tests

File: tests/test_expectation_maximization.py [hunk 4]
```diff
@@ -23,24 +23,40 @@ def n_particles():
...
-        "bf": 0.1,
-        "lp": 0.1,
+    pixels = n_pix
+    ctf_info = {
+        "amplitude_contrast": 0.1,
+        "b_factor": 0.0,
+        "batch_size": n_particles,
+        "cs": 2.7,
+        "ctf_size": pixels,
+        "kv": 0.1968,
```
Comment (geoffwoollard) [inline tests/test_expectation_maximization.py]: kv could/should be 300. in that case wavelength would be 0.1968.

it's my understanding that `ctf.get_ctf` needs `wavelength` and doesn't use `kv`

**Review by geoffwoollard — CHANGES_REQUESTED** [review]
Close... but a bit more work needed.

I like the tests... how they have the `n_particles` for defocus etc and work with `test_ir`

**Review by geoffwoollard — APPROVED** [review]
Would be good to show screen shot of what the ctf looks like with these values as sanity check. I'll merge, we'll see downstream.

**Comment by geoffwoollard** [comment]
Fixes https://github.com/compSPI/reconstructSPI/issues/11
